File: packages/flutter_foreground_task/lib/models/notification_icon.dart
import 'dart:ui';

import 'package:flutter_foreground_task/utils/color_extension.dart';

/// A data class for dynamically changing the notification icon.
class NotificationIcon {
  /// Constructs an instance of [NotificationIcon].
  const NotificationIcon({
    this.metaDataName = '',
    this.backgroundColor,
    this.largeIconPath,
  });

  /// The name of the meta-data in the manifest that contains the drawable icon resource identifier.
  /// Leave empty to keep the app's default small icon.
  final String metaDataName;

  /// The background color for the notification icon.
  final Color? backgroundColor;

  /// Absolute path to a local image file to show as the notification's large
  /// icon (e.g. a photo relevant to what the notification is about).
  ///
  /// Fork-only addition (not part of upstream `flutter_foreground_task`,
  /// which has no way to set a bitmap large icon — only a build-time bundled
  /// small-icon drawable via [metaDataName]).
  final String? largeIconPath;

  /// Returns the data fields of [NotificationIcon] in JSON format.
  Map<String, dynamic> toJson() {
    return {
      'metaDataName': metaDataName,
      'backgroundColorRgb': backgroundColor?.toRgbString,
      'largeIconPath': largeIconPath,
    };
  }

  /// Creates a copy of the object replaced with new values.
  NotificationIcon copyWith({
    String? metaDataName,
    Color? backgroundColor,
    String? largeIconPath,
  }) =>
      NotificationIcon(
        metaDataName: metaDataName ?? this.metaDataName,
        backgroundColor: backgroundColor ?? this.backgroundColor,
        largeIconPath: largeIconPath ?? this.largeIconPath,
      );
}

File: packages/flutter_foreground_task/android/src/main/kotlin/com/pravera/flutter_foreground_task/models/NotificationIcon.kt
package com.pravera.flutter_foreground_task.models

import org.json.JSONObject

data class NotificationIcon(
    val metaDataName: String,
    val backgroundColorRgb: String?,
    // Fork-only addition: absolute path to a local image file to show as the
    // notification's large icon (not part of upstream flutter_foreground_task).
    val largeIconPath: String? = null
) {
    companion object {
        private const val META_DATA_NAME_KEY = "metaDataName"
        private const val BACKGROUND_COLOR_RGB_KEY = "backgroundColorRgb"
        private const val LARGE_ICON_PATH_KEY = "largeIconPath"

        fun fromJsonString(jsonString: String): NotificationIcon {
            val jsonObj = JSONObject(jsonString)

            val metaDataName: String = if (jsonObj.isNull(META_DATA_NAME_KEY)) {
                ""
            } else {
                jsonObj.getString(META_DATA_NAME_KEY)
            }

            val backgroundColorRgb: String? = if (jsonObj.isNull(BACKGROUND_COLOR_RGB_KEY)) {
                null
            } else {
                jsonObj.getString(BACKGROUND_COLOR_RGB_KEY)
            }

            val largeIconPath: String? = if (!jsonObj.has(LARGE_ICON_PATH_KEY) || jsonObj.isNull(LARGE_ICON_PATH_KEY)) {
                null
            } else {
                jsonObj.getString(LARGE_ICON_PATH_KEY)
            }

            return NotificationIcon(
                metaDataName = metaDataName,
                backgroundColorRgb = backgroundColorRgb,
                largeIconPath = largeIconPath
            )
        }
    }
}
